tower/src/util/pending.rs

```diff
@@ -0,0 +1,141 @@
+use std::fmt;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_service::Service;
+
+/// Returns a new `Pending` for the given future.
+///
+/// # Example
+/// ```
+/// use tower::{service_fn, Service, ServiceExt};
+/// use tower::util::pending;
+/// use std::convert::Infallible;
+///
+/// # fn main() {
+/// # async {
+/// let mut pending_svc = pending(Box::pin(async {
+///     let svc = service_fn(|_req: ()| async { Ok::<_, Infallible>("ok") });
+///     Ok::<_, Infallible>(svc)
+/// }));
+///
+/// let svc = pending_svc.ready_and().await.unwrap();
+/// let res = svc.call(()).await.unwrap();
+/// # };
+/// # }
+/// ```
+pub fn pending<F, S, R, E>(future: F) -> Pending<F, S>
+where
+    F: Future<Output = Result<S, E>> + Unpin,
+    S: Service<R, Error = E>,
+{
+    Pending {
+        inner: Inner::Future(future),
+    }
+}
+
+/// A type that implements `Service` for a `Future` that produces a `Service`.
+#[derive(Clone)]
+pub struct Pending<F, S> {
+    inner: Inner<F, S>,
+}
+
+impl<F, S> fmt::Debug for Pending<F, S>
+where
+    S: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        #[derive(Debug)]
+        struct Pending;
+
+        let mut f = f.debug_struct("Pending");
+        let f = match &self.inner {
+            Inner::Future(_) => f.field("service", &Pending),
+            Inner::Service(svc) => f.field("service", svc),
+        };
```

Comment:
take it or leave it: I have a slight preference for formatting this a little bit more faithfully to its actual implementation, so that the enum is preserved. for example, when pending, it could look like
```
Pending { inner: Inner::Future(...) }
```
and when complete,
```
Pending { inner: Inner::Service(MyService { whatever }) }
```

Reply:
👌 Changed in 77fce4df59f16